### DISCUSSION TITLE
Audio Crackling For Flash Memory Playback

### DISCUSSION THREAD

**Discussion by BB-check** [pschatzmann/arduino-audio-tools#542] [Q&A] [answered]
I've been trying to integrate sound into certain projects without additional hardware which is why I've been looking into putting small sounds (for sound effect feedback on button press) entirely into memory. For reference, when I used the "Star Wars" example with my setup (MAX98357 amp with an 8ohm speaker) the audio has no crackling. However, when I try to play short sound effects that I cut and load into an array of memory streams as a sort of audio reference table, I get crackling towards the end of the sound effect. 

In audacity, when cut out the relevant sound effects I ensure that the file is mono and not stereo and of the sample rate 44100Khz, and in playback on my laptop, I don't hear any crackling. I then encode it into a C style array using a hex editor like the example. However on playback with the test sketch enclosed, the audio starts to crackle/have a fuzz near the end of playback. I'm not sure if its the fact that I'm cutting the audio and there's some sort of artifacting effect happening or something to do with code handling the end of the byte array.

`
#include "AudioTools.h"
#include "StarWars30.h"
#include "click.h"

uint8_t channels = 1;
uint16_t sample_rate = 44100;

I2SStream i2s;  // Output to I2S
//MemoryStream music(StarWars30_raw, StarWars30_raw_len);
MemoryStream soundEffects[] = { MemoryStream (gurgleSoundOneData, gurgleSoundOneDataLength), MemoryStream (gurgleSoundTwoData, gurgleSoundTwoDataLength), MemoryStream (gurgleSoundThreeData, gurgleSoundThreeDataLength), MemoryStream (gurgleSoundFourData, gurgleSoundFourDataLength)};

StreamCopyT<int16_t> copier(i2s, soundEffects[0]); // copies sound into i2s

void setup(){
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);
    auto config = i2s.defaultConfig(TX_MODE);
    config.sample_rate = sample_rate;
    config.channels = channels;
    config.bits_per_sample = 16;
    i2s.begin(config);
}

//blocking playback audio function
void playAudio(int selectedAudio){
 
 long beginning = millis();
  copier.begin(i2s, soundEffects[selectedAudio]);
  while (copier.copy()){
    //block program execution while memory is copied to buffer.
  }

  //reset memory stream object to play from beginning 
  soundEffects[selectedAudio].begin();

}

void loop(){
  
 for (int i = 0; i < 4; i++){
    playAudio(i);
    delay(1500);//give time to let audio play completely before next sound
 }

}`

**Comment by pschatzmann** [comment, marked as answer]
My guess is that I2S does not like it if you abruptly stop to provide any data. I suggest that you check what happens if you add plenty of silence at the end before you stop.
ps. instead of your while(copier.copy()) loop you can just use a copier.copyAll().

**Reply by BB-check** [reply]
So looking at the waveforms in audacity, yes the audio does cut abruptly. When there is no silence padding for the track, you get the crackling as I described above. When I add about a 200ms tail of silence, there's no more crackling but there is an audible pop at the end of playback of a sound effect. With 300 ms of padding, there's no noise at the end of playback. 

Oddly enough, when I decided to increase the number of buffers allocated for the config object from 10 ( I think the default) to 20, the audible popping came back for the audio tracks even with 300 ms of padding.

**Comment by joba-1** [comment]
ESPs do some stuff in the background and need some spare CPU cycles for this.
I think the while loop needs a delay() of some sort.
Not sure this helps with your problem though.

**Reply by zachtos** [reply]
I'm doing the exact same issue.  Clicks/pops at end of files.  Same setup, esp32, 16mB flash, reading from flash, playing as WAV 22khz mono, max98357a, i2s.  Sounds great until the end of the file. I have similar problems with earlephilhower library but there were several work arounds.  Are there any solutions for this that do not require a huge silence at file end?  I open audio files rapidly for menu type work.  Other library had a audio.flush() function that consumes all samples before ending to prevent the pop.

also this library seems incredibly difficult to do a simple task, such as open a file by name, where name is a variable and not fixed.

**Comment by pschatzmann** [comment]
I don't understand your comment regarding files: Just open the file and copy the content to the output stream. 
What is so difficult about this ? 
See [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd-audiokit/streams-sd-audiokit.ino)
